### DISCUSSION TITLE
Flatpickr locale

### DISCUSSION THREAD

**Discussion by psjavi** [rappasoft/laravel-livewire-tables#1739] [Q&A]
Hi, I need to set the locale to "ES" for the Flatpickr range filter.

How can I do it? I've tried to don't inyect the package's assets and install it via npm but I don't know how to I have to init the flatpicker declared into de package.

Thanks in advance.

**Comment by lrljoe** [comment]
I will add this in when I add in the defaultDate for the range filter, hopefully in the next week.

**Reply by psjavi** [reply]
Ok, thanks a lot

**Reply by lrljoe** [reply]
Please reach out on Discord if you'd like an early access branch to test with, as I'm in the process of updating the docs for this.

**Reply by lrljoe** [reply]
Merged into development, if you'd like to test

**Reply by psjavi** [reply]
Hi, thanks a lot.

I've been reading the docs but I'm not able to find out how I have to change the locale.

I've got my .env config with `APP_LOCALE=es
`
The date range x-data created in the html is the following:

`x-data="flatpickrFilter($wire, 'fecha_de_solicitud', JSON.parse('{"allowInput":true,"altFormat":"F j, Y","ariaDateFormat":"F j, Y","dateFormat":"d/m/Y","earliestDate":null,"latestDate":null}'), $refs.dateRangeInput, 'es')">`

My php is the following:

```
            DateRangeFilter::make('Fecha de solicitud')
                ->config([
                    'dateFormat' => 'd/m/Y'
                ])
                ->setFilterPillValues([0 => 'minDate', 1 => 'maxDate']) // The values that will be displayed for the Min/Max Date Values
                ->filter(function (Builder $builder, array $dateRange) { // Expects an array.
                    $builder
                        ->whereDate('requests.created_at', '>=', $dateRange['minDate']) // minDate is the start date selected
                        ->whereDate('requests.created_at', '<=', $dateRange['maxDate']); // maxDate is the end date selected
                }),
```

I've also tried to put in the php this configuration:

```
            DateRangeFilter::make('Fecha de solicitud')
                ->config([
                    'dateFormat' => 'd/m/Y',
'locale' => 'es'
                ])
```
What I'm doing wrong?

Thanks a lot in advance.

**Reply by lrljoe** [reply]
So, if you want the Spanish locale, then you need to ensure that the locale is loaded.

The docs aren't updated yet, but the relevant content is:

## Localisation (BETA)
The default installation includes only the English (en) locale.

### Bundling
Should you wish to localise, you must include the Flatpickr locale files in your build pipeline.  This applies to only the specific locales that you require in your app.js (requires adding the flatpickr library to your package.json by executing "npm i flatpickr --save")

```
import { Spanish } from "../imports/flatpickr/l10n/es.js";
```

### CDN
You can also add locales using the Flatpickr CDN, ensuring that these are loaded before the page renders.

For example to add Spanish (es), ensure that the following is in the "head" section of your layout, ideally before your app.js
```
<script src="https://npmcdn.com/flatpickr/dist/l10n/es.js" async></script>
```

**Reply by lrljoe** [reply]
Docs should be updated in next day or two to reflect the above.

**Reply by psjavi** [reply]
Good morning Joe,

Sorry for disturbing you. I've done some test but the calendar is still in english.

![image](https://github.com/user-attachments/assets/ccc1f91f-1553-49ca-aaf2-b737efe1fc2a)

I've installed flatpicker as package:

![image](https://github.com/user-attachments/assets/f57067e7-1c15-4c01-a125-91a3266df97b)

I've added the locale in my app.js. I've also done the following attempt:

![image](https://github.com/user-attachments/assets/5f8a5b75-323c-484c-9202-3bcdbbbddfd5)

The app.js is correctly compiled with the locale:

![image](https://github.com/user-attachments/assets/c820c0bd-6698-4a8f-97f4-19fa844341b1)

I've exported the date-range view in order to modify the html but I also have tried to restore it and use the view comming form vendor. This is my custom html:

```
@php
    $filterKey = $filter->getKey();
@endphp

<div x-cloak id="{{ $tableName }}-dateRangeFilter-{{ $filterKey }}"
     x-data="flatpickrFilter($wire, '{{ $filterKey }}', @js($filter->getConfigs()), $refs.dateRangeInput, '{{ App::currentLocale() }}')">

    <div class="relative z-0">
        <input
            type="text"
            x-ref="dateRangeInput"
            x-on:click="init"
            value="{{ $filter->getDateString(isset($this->appliedFilters[$filterKey]) ? $this->appliedFilters[$filterKey] : '') }}"
            wire:key="{{ $filter->generateWireKey($tableName, 'dateRange') }}"
            id="{{ $tableName }}-filter-dateRange-{{ $filterKey }}"
            class="pt-3 pb-2 block w-full px-0 mt-0 bg-transparent border-0 border-b-2 appearance-none focus:outline-none focus:ring-0 focus:border-black border-gray-200"
            placeholder=" "
        />
        <x-livewire-tables::tools.filter-label :$filter :$tableName />
    </div>
</div>
```

What I'm doing wrong?

Many thanks for your help and time.

**Comment by ADRYEP** [comment]
Hi! I'm having the same error. @psjavi did you solve it?

**Reply by lrljoe** [reply]
@ADRYEP @psjavi 

Can you both share the code for your DateRangeFilter from your Table please, and confirm the version of the Tables you're using, 
and 

Whether you've previously published the views (recommend not doing this!)

and

Try running a fresh "npm run build" and "php artisan view:clear" please

**Reply by ADRYEP** [reply]
Hi,

This is the table component file chunk where I create the Date filter:

            DateRangeFilter::make('Fecha de solicitud')
                ->config([
                    'dateFormat' => 'd/m/Y'
                ])
                ->setFilterPillValues([0 => 'minDate', 1 => 'maxDate']) // The values that will be displayed for the Min/Max Date Values
                ->filter(function (Builder $builder, array $dateRange) { // Expects an array.
                    $builder
                        ->whereDate('requests.created_at', '>=', $dateRange['minDate']) // minDate is the start date selected
                        ->whereDate('requests.created_at', '<=', $dateRange['maxDate']); // maxDate is the end date selected
                })
                
Now this is the blade template i'm using:

@php
    $filterKey = $filter->getKey();
@endphp

<div x-cloak id="{{ $tableName }}-dateRangeFilter-{{ $filterKey }}"
     x-data="flatpickrFilter($wire, '{{ $filterKey }}', @js($filter->getConfigs()), $refs.dateRangeInput, '{{ App::currentLocale() }}')">

    <div class="relative z-0">
        <input
            type="text"
            x-ref="dateRangeInput"
            x-on:click="init"
            value="{{ $filter->getDateString(isset($this->appliedFilters[$filterKey]) ? $this->appliedFilters[$filterKey] : '') }}"
            wire:key="{{ $filter->generateWireKey($tableName, 'dateRange') }}"
            id="{{ $tableName }}-filter-dateRange-{{ $filterKey }}"
            class="pt-3 pb-2 block w-full px-0 mt-0 bg-transparent border-0 border-b-2 appearance-none focus:outline-none focus:ring-0 focus:border-black border-gray-200"
            placeholder=" "
        />
        <x-livewire-tables::tools.filter-label :$filter :$tableName />
    </div>
</div>

I'm using 3.3 Tables version and I did published some of the views I needed to customize. Although, I erased thos and did a clean installation of the Tables packages and I was not able to change the Flatpickr language.

![Captura de pantalla 2024-08-01 103110](https://github.com/user-attachments/assets/a3a76ab9-18ad-4a7c-ae7f-3ad95d424026)

I have done several times npm run build/dev and php artisan view:clear

Thanks a lot for the suppor

**Reply by lrljoe** [reply]
Ah, this should be an easy one to fix, you're missing the "locale" option from the config section:
```
        DateRangeFilter::make('Fecha de solicitud')
            ->config([
                'dateFormat' => 'd/m/Y',
               'locale' => 'es'
            ])
```
It uses that to determine which locale to use, rather than any globally defined flatpickr localisation.

**Reply by ADRYEP** [reply]
Hi!
This didn't fix the issue.

![image](https://github.com/user-attachments/assets/888ea5d9-ea2b-413b-b27f-16b9ba761408)

![image](https://github.com/user-attachments/assets/9a0c94c9-7b23-40e4-a74d-20fda1bdc3f0)

**Reply by lrljoe** [reply]
@ADRYEP how are you adding the localisations for Flatpickr? Are they bundled in, or are you using a CDN to add them to your layout?

**Reply by ADRYEP** [reply]
None of both, I directly use it in the Table Component and the blade:

blade:
@php
    $filterKey = $filter->getKey();
@endphp

<div x-cloak id="{{ $tableName }}-dateRangeFilter-{{ $filterKey }}"
     x-data="flatpickrFilter($wire, '{{ $filterKey }}', @js($filter->getConfigs()), $refs.dateRangeInput, '{{ App::currentLocale() }}')">

    <div class="relative z-0">
        <input
            type="text"
            x-ref="dateRangeInput"
            x-on:click="init"
            value="{{ $filter->getDateString(isset($this->appliedFilters[$filterKey]) ? $this->appliedFilters[$filterKey] : '') }}"
            wire:key="{{ $filter->generateWireKey($tableName, 'dateRange') }}"
            id="{{ $tableName }}-filter-dateRange-{{ $filterKey }}"
            class="pt-3 pb-2 block w-full px-0 mt-0 bg-transparent border-0 border-b-2 appearance-none focus:outline-none focus:ring-0 focus:border-black border-gray-200"
            placeholder=" "
        />
        <x-livewire-tables::tools.filter-label :$filter :$tableName />
    </div>
</div>


Livewire component:
            DateRangeFilter::make('Fecha de solicitud')
                ->config([
                    'dateFormat' => 'd/m/Y'
                ])
                ->setFilterPillValues([0 => 'minDate', 1 => 'maxDate']) // The values that will be displayed for the Min/Max Date Values
                ->filter(function (Builder $builder, array $dateRange) { // Expects an array.
                    $builder
                        ->whereDate('requests.created_at', '>=', $dateRange['minDate']) // minDate is the start date selected
                        ->whereDate('requests.created_at', '<=', $dateRange['maxDate']); // maxDate is the end date selected
                })

**Reply by lrljoe** [reply]
> None of both, I directly use it in the Table Component and the blade:
> 
> blade:
> @php
>     $filterKey = $filter->getKey();
> @endphp
> 
> <div x-cloak id="{{ $tableName }}-dateRangeFilter-{{ $filterKey }}"
>      x-data="flatpickrFilter($wire, '{{ $filterKey }}', @js($filter->getConfigs()), $refs.dateRangeInput, '{{ App::currentLocale() }}')">
> 
>     <div class="relative z-0">
>         <input
>             type="text"
>             x-ref="dateRangeInput"
>             x-on:click="init"
>             value="{{ $filter->getDateString(isset($this->appliedFilters[$filterKey]) ? $this->appliedFilters[$filterKey] : '') }}"
>             wire:key="{{ $filter->generateWireKey($tableName, 'dateRange') }}"
>             id="{{ $tableName }}-filter-dateRange-{{ $filterKey }}"
>             class="pt-3 pb-2 block w-full px-0 mt-0 bg-transparent border-0 border-b-2 appearance-none focus:outline-none focus:ring-0 focus:border-black border-gray-200"
>             placeholder=" "
>         />
>         <x-livewire-tables::tools.filter-label :$filter :$tableName />
>     </div>
> </div>
> 
> 
> Livewire component:
>             DateRangeFilter::make('Fecha de solicitud')
>                 ->config([
>                     'dateFormat' => 'd/m/Y'
>                 ])
>                 ->setFilterPillValues([0 => 'minDate', 1 => 'maxDate']) // The values that will be displayed for the Min/Max Date Values
>                 ->filter(function (Builder $builder, array $dateRange) { // Expects an array.
>                     $builder
>                         ->whereDate('requests.created_at', '>=', $dateRange['minDate']) // minDate is the start date selected
>                         ->whereDate('requests.created_at', '<=', $dateRange['maxDate']); // maxDate is the end date selected
>                 })
> 

Looks like you're missing the locale key

How are you adding the Flatpickr es translation, you need to either bundle it, or use a cdn to add the translation file for flatpickr, which one are you using.

**Reply by ADRYEP** [reply]
Could you show me an example bundling it? I tried using the translation file via CDN but it didn't work and I tried specifying the translation in the Laravel proper file but it didn't work either

**Reply by lrljoe** [reply]
@ADRYEP  - please read the docs for the DateRangeFilter, it should be clear from those! 

IF you still have a problem - please share your 

1. app.js
2. The "head" section of the relevant layout
3. The filters() section of your table.

**Comment by lrljoe** [comment]
If anyone is still struggling with this, please do let me know, reach out on the Discord (ideally) as the best way to reach me, and I'll try to think of a way to document this more clearly.
